## homework 4 - Functions II

1. Import the data as a pd.DataFrame object. Use one or more functions from the pandas library to split the data into three unique datasets. The first should include columns that are factors only (i.e., categorical data), the second should include columns that are numeric only, and the third should include columns with logical/boolean variables only.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("Housing_prices_data.csv")

In [3]:
df.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [4]:
df_categorical = df.select_dtypes(include=['object', 'category'])

In [5]:
df_numeric = df.select_dtypes(include=['number'])

In [6]:
df_boolean = df.select_dtypes(include=['bool'])

In [7]:
print("Categorical columns:", df_categorical.shape)
print("Numeric columns:", df_numeric.shape)
print("Boolean columns:", df_boolean.shape)

Categorical columns: (1460, 42)
Numeric columns: (1460, 38)
Boolean columns: (1460, 1)


In [8]:
df_categorical.head()

,MSZoning,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,...,GarageType,GarageFinish,GarageQual,GarageCond,PavedDrive,PoolQC,Fence,MiscFeature,SaleType,SaleCondition
0,RL,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,...,Attchd,RFn,TA,TA,Y,NaN,NaN,NaN,WD,Normal
1,RL,Pave,NaN,Reg,Lvl,AllPub,FR2,Gtl,Veenker,Feedr,...,Attchd,RFn,TA,TA,Y,NaN,NaN,NaN,WD,Normal
2,RL,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,...,Attchd,RFn,TA,TA,Y,NaN,NaN,NaN,WD,Normal
3,RL,Pave,NaN,IR1,Lvl,AllPub,Corner,Gtl,Crawfor,Norm,...,Detchd,Unf,TA,TA,Y,NaN,NaN,NaN,WD,Abnorml
4,RL,Pave,NaN,IR1,Lvl,AllPub,FR2,Gtl,NoRidge,Norm,...,Attchd,RFn,TA,TA,Y,NaN,NaN,NaN,WD,Normal


df_numeric.head()

In [9]:
df_boolean.head()

,CentralAir
0,True
1,True
2,True
3,True
4,True


2. Using the second dataset from question #1, for each numeric variable use the apply function from the pandas library to return the mean, median and standard deviation (SD) of all numeric variables.

In [10]:
summary_stats = df_numeric.apply(lambda x: pd.Series({"mean": x.mean(),"median": x.median(),"sd": x.std()}))
print(summary_stats)

                Id  MSSubClass  LotFrontage       LotArea  OverallQual  \
mean    730.500000   56.897260    70.049958  10516.828082     6.099315   
median  730.500000   50.000000    69.000000   9478.500000     6.000000   
sd      421.610009   42.300571    24.284752   9981.264932     1.382997   

        OverallCond    YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  \
mean       5.575342  1971.267808   1984.865753  103.685262  443.639726  ...   
median     5.000000  1973.000000   1994.000000    0.000000  383.500000  ...   
sd         1.112799    30.202904     20.645407  181.066207  456.098091  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  \
mean     94.244521    46.660274      21.954110   3.409589    15.060959   
median    0.000000    25.000000       0.000000   0.000000     0.000000   
sd      125.338794    66.256028      61.119149  29.317331    55.757415   

         PoolArea     MiscVal    MoSold       YrSold      SalePrice  
mean     2.758904  

3. Using the second dataset from question #1, create a new dataset that only includes the variable indicating sales price ("SalePrice"). Search for a categorical variable which has between 2-4 categories in the full dataset that you think may be related to the sales price of houses. Give the mean, median and SD for SalePrice based on the different groups of the categorical variables. Does the average price seem to vary by the different levels of the categorical factor you chose?

In [11]:
price_drive = df[['SalePrice', 'PavedDrive']]

In [12]:
price_stats = price_drive.groupby('PavedDrive')['SalePrice'].apply(lambda x: pd.Series({'mean': x.mean(),'median': x.median(),'sd': x.std()}))

print(price_stats)

PavedDrive        
N           mean      115039.122222
            median    111000.000000
            sd         44352.523309
P           mean      132330.000000
            median    132250.000000
            sd         33503.030228
Y           mean      186433.973881
            median    168500.000000
            sd         79665.503047
Name: SalePrice, dtype: float64


Yes. Houses with paved driveways (Y) have the highest mean sale price (about $186,434) and median price ($168,500). Homes with partially paved driveways (P) have moderately lower prices, while homes with unpaved driveways (N) have the lowest mean and median sale prices.

4. Regress "SalePrice" on "LotArea", "OverallQual", "OverallCond" with the statsmodels library. Use the syntax below to replicate the output shown you are able to estimate this regression model.

In [13]:
import statsmodels.api as sm

In [14]:
x = df_numeric[['LotArea','OverallQual', 'OverallCond']]
y = (df_numeric["SalePrice"])
x = sm.add_constant(x)

In [15]:
model = sm.OLS(y,x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:              SalePrice   R-squared:                       0.659
Model:                            OLS   Adj. R-squared:                  0.658
Method:                 Least Squares   F-statistic:                     935.9
Date:                Thu, 05 Mar 2026   Prob (F-statistic):               0.00
Time:                        12:12:57   Log-Likelihood:                -17760.
No. Observations:                1460   AIC:                         3.553e+04
Df Residuals:                    1456   BIC:                         3.555e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.022e+05   8633.073    -11.832      

5. Write a for loop that loops through each level of the categorical factor you chose in Q3, where each iteration of the for loop subsets your data for one group of the categorical variable at a time, and then fits the same regression model specified above in Q4 k times, where k refers to the number of unique categories. On each iteration of your for loop, store the model.summary() information as a new element in a list object. Once your for loop is done running, print the list. There should be k regressions which you estimated. Does the effect of LotArea on SalePrice vary within the different levels of the categorical factor you chose? What is another, more common way to answer a question like this (i.e., a question about moderation or effect heterogeneity)? Hint: Don't forget to account for the standard error when comparing the regression slopes!

In [16]:
model_list = []

In [17]:
levels = df['PavedDrive'].unique()
for level in levels:
    subset = df[df['PavedDrive'] == level]
    x = subset[['LotArea','OverallQual','OverallCond']]
    y = subset['SalePrice']
    x = sm.add_constant(x)
    model = sm.OLS(y, x).fit()
    model_list.append(model.summary())

print(model_list)

[<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
Dep. Variable:              SalePrice   R-squared:                       0.657
Model:                            OLS   Adj. R-squared:                  0.656
Method:                 Least Squares   F-statistic:                     853.3
Date:                Thu, 05 Mar 2026   Prob (F-statistic):          7.87e-310
Time:                        12:12:57   Log-Likelihood:                -16307.
No. Observations:                1340   AIC:                         3.262e+04
Df Residuals:                    1336   BIC:                         3.264e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
c

The estimated coefficient of LotArea differs across the levels of PavedDrive, suggesting that the relationship between lot size and sale price may vary by driveway type. However, the standard errors for the smaller groups are relatively large, so the differences in slopes may not be statistically significant. A more common way to test for this type of effect heterogeneity is to include interaction terms between LotArea and PavedDrive in a single regression model.

6. Using the first dataset from question #1, for each categorical variable use a for loop to print a frequency and relative frequency table for every variable. To make your output readable, for each variable you should print the following in this order, placing the variable name where you see blanks below.

In [18]:
for col in df_categorical.columns:
    
    print(f"Frequency table for {col} ============")
    print(df_categorical[col].value_counts())
    print()
    print(f"Relative frequency table for {col} ============")
    print(df_categorical[col].value_counts(normalize=True))
    print("\n\n")

Frequency table for MSZoning ============
MSZoning
RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: count, dtype: int64

Relative frequency table for MSZoning ============
MSZoning
RL         0.788356
RM         0.149315
FV         0.044521
RH         0.010959
C (all)    0.006849
Name: proportion, dtype: float64



Frequency table for Street ============
Street
Pave    1454
Grvl       6
Name: count, dtype: int64

Relative frequency table for Street ============
Street
Pave    0.99589
Grvl    0.00411
Name: proportion, dtype: float64



Frequency table for Alley ============
Alley
Grvl    50
Pave    41
Name: count, dtype: int64

Relative frequency table for Alley ============
Alley
Grvl    0.549451
Pave    0.450549
Name: proportion, dtype: float64



Frequency table for LotShape ============
LotShape
Reg    925
IR1    484
IR2     41
IR3     10
Name: count, dtype: int64

Relative frequency table for LotShape ============
LotShape
Reg    0.633562
IR1   

7. Do the same thing as number 6, but with a while loop this time.

In [19]:
cols = df_categorical.columns
i = 0

while i < len(cols):
    col = cols[i]
    print(f"Frequency table for {col} ============")
    print(df_categorical[col].value_counts())
    print()
    print(f"Relative frequency table for {col} ============")
    print(df_categorical[col].value_counts(normalize=True))
    print("\n\n")
    i += 1

Frequency table for MSZoning ============
MSZoning
RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: count, dtype: int64

Relative frequency table for MSZoning ============
MSZoning
RL         0.788356
RM         0.149315
FV         0.044521
RH         0.010959
C (all)    0.006849
Name: proportion, dtype: float64



Frequency table for Street ============
Street
Pave    1454
Grvl       6
Name: count, dtype: int64

Relative frequency table for Street ============
Street
Pave    0.99589
Grvl    0.00411
Name: proportion, dtype: float64



Frequency table for Alley ============
Alley
Grvl    50
Pave    41
Name: count, dtype: int64

Relative frequency table for Alley ============
Alley
Grvl    0.549451
Pave    0.450549
Name: proportion, dtype: float64



Frequency table for LotShape ============
LotShape
Reg    925
IR1    484
IR2     41
IR3     10
Name: count, dtype: int64

Relative frequency table for LotShape ============
LotShape
Reg    0.633562
IR1   

8. Using the Rocket class which we created as part of the week 7 class activity, we are going to create a new class called shuttle. The shuttle class should inherit all of Rocket's characteristics but has one additional parameter: flights_completed which measures the number of flights the shuttle has completed.

In [20]:
from math import sqrt

class Rocket():
    """Rocket simulates a rocket ship for a game or a physics simulation."""
    
    def __init__(self, x=0, y=0):
        """Each rocket has an (x,y) position in 2D space."""
        self.x = x
        self.y = y
        
    def move_rocket(self, x_increment=0, y_increment=1):
        """Moves the rocket according to the paremeters given.
           Default behavior is to move the rocket up one unit."""
        self.x += x_increment
        self.y += y_increment
        
    def get_distance(self, other_rocket):
        """Calculates the distance from this rocket to another rocket,
           and returns that value"""
        distance = sqrt((self.x-other_rocket.x)**2+(self.y-other_rocket.y)**2)
        return distance
    

In [21]:
class Shuttle(Rocket):
    
    def __init__(self, x=0, y=0, flights_completed=0):
        super().__init__(x, y)
        self.flights_completed = flights_completed

my_shuttle = Shuttle(x=10, y=25, flights_completed=7)

print(f"The shuttle is located at ({my_shuttle.x}, {my_shuttle.y}), and has recorded {my_shuttle.flights_completed} flights")

The shuttle is located at (10, 25), and has recorded 7 flights


9. Write a single for loop which randomly creates 11 rockets and 11 shuttles using randomly generated integers. Store this output in two separate lists. Once you have your data, use for loops to print the messages you see below. Note you can use the randint() function from the random module (which you can import with the command from random import randint) to randomly generate the characteristics of your Rockets and Shuttles.

In [22]:
from random import randint
from math import sqrt
rockets = []
shuttles = []

In [23]:
for _ in range(11):
    rocket_x = randint(0, 100)
    rocket_y = randint(0, 100)
    rockets.append(Rocket(x=rocket_x, y=rocket_y))
    shuttle_x = randint(0, 100)
    shuttle_y = randint(0, 100)
    flights = randint(0, 20)
    shuttles.append(Shuttle(x=shuttle_x, y=shuttle_y, flights_completed=flights))

In [24]:
for idx, shuttle in enumerate(shuttles, start=1):
    print(f"Shuttle {idx} has completed {shuttle.flights_completed} flights.")

Shuttle 1 has completed 17 flights.
Shuttle 2 has completed 10 flights.
Shuttle 3 has completed 9 flights.
Shuttle 4 has completed 17 flights.
Shuttle 5 has completed 17 flights.
Shuttle 6 has completed 4 flights.
Shuttle 7 has completed 6 flights.
Shuttle 8 has completed 18 flights.
Shuttle 9 has completed 4 flights.
Shuttle 10 has completed 12 flights.
Shuttle 11 has completed 18 flights.


In [25]:
for i in range(1, len(shuttles)):
    distance = shuttles[i].get_distance(shuttles[i-1])
    print(f"Shuttle {i+1} is {distance:.2f} units away from Shuttle {i}.")

Shuttle 2 is 31.58 units away from Shuttle 1.
Shuttle 3 is 45.54 units away from Shuttle 2.
Shuttle 4 is 52.95 units away from Shuttle 3.
Shuttle 5 is 28.30 units away from Shuttle 4.
Shuttle 6 is 49.52 units away from Shuttle 5.
Shuttle 7 is 47.51 units away from Shuttle 6.
Shuttle 8 is 73.16 units away from Shuttle 7.
Shuttle 9 is 30.41 units away from Shuttle 8.
Shuttle 10 is 19.85 units away from Shuttle 9.
Shuttle 11 is 63.56 units away from Shuttle 10.


In [26]:
for i in range(len(shuttles)):
    distance = shuttles[i].get_distance(rockets[i])
    print(f"Shuttle {i+1} is {distance:.2f} units away from Rocket {i+1}.")

Shuttle 1 is 44.20 units away from Rocket 1.
Shuttle 2 is 61.85 units away from Rocket 2.
Shuttle 3 is 93.19 units away from Rocket 3.
Shuttle 4 is 60.83 units away from Rocket 4.
Shuttle 5 is 34.18 units away from Rocket 5.
Shuttle 6 is 25.32 units away from Rocket 6.
Shuttle 7 is 79.61 units away from Rocket 7.
Shuttle 8 is 75.33 units away from Rocket 8.
Shuttle 9 is 33.24 units away from Rocket 9.
Shuttle 10 is 51.00 units away from Rocket 10.
Shuttle 11 is 65.62 units away from Rocket 11.
